### PR TITLE
Relocation of DW_OP_{bra, skip} instructions

### PR DESCRIPTION
**Pull request by ggreif**
The fundamental problem is that the target distance of jump-like operations may change in the DWARF expression translation process. Intervening `DW_OP_deref` will expand to about 10 bytes, for example.

So the jumps must be relocated. We approach this task by inserting artificial `LandingPad` markers (new `CompiledExpressionParts` constructors) into the parsed vector at actual `Jump` targets.

`LandingPad`s are identified by `JumpTargetMarker` tokens which are generated on the fly.

Additionally we now parse the `Jump` instructions. These also get their corresponding `JumpTargetMarker` token.

We bail in two situations:
- `frame_base` is too complicated (i.e. itself contains `Jump`)
- some jump distance in the original expression is fishy.

Open problem: the `need_deref` flag. What when different branches want to set the flag differently? Probably not a biggie as long as all control-flow paths end e.g. with `DW_OP_stack_value` (or without).

---------------
Example:
``` haskell
DW_AT_location	(DW_OP_WASM_location 0x0 +1, DW_OP_dup, DW_OP_lit1, DW_OP_and, DW_OP_bra +5, DW_OP_lit1, DW_OP_shr, DW_OP_skip +3, DW_OP_plus_uconst 0x5, DW_OP_deref, DW_OP_stack_value)
```
gets transformed into:
``` haskell
DW_AT_location	(DW_OP_breg6 RBP-36, DW_OP_deref, DW_OP_dup, DW_OP_lit1, DW_OP_and, DW_OP_bra +5, DW_OP_lit1, DW_OP_shr, DW_OP_skip +19, DW_OP_plus_uconst 0x5, DW_OP_breg6 RBP-56, DW_OP_deref, DW_OP_consts +152, DW_OP_plus, DW_OP_deref, DW_OP_swap, DW_OP_const4u 0xffffffff, DW_OP_and, DW_OP_plus, DW_OP_deref, DW_OP_stack_value)
```

### CODE REVIEW

File: crates/debug/src/transform/expression.rs [hunk 1]
```diff
@@ -87,6 +87,10 @@ enum CompiledExpressionPart {
     Local { label: ValueLabel, trailing: bool },
     // Dereference is needed.
     Deref,
+    // Jumping in the expression.
+    Jump { target: i16, conditionally: bool },
```
Comment (ggreif) [inline crates/debug/src/transform/expression.rs:91]: `target` is informational only, never used

File: crates/debug/src/transform/expression.rs [hunk 2]
```diff
@@ -346,6 +381,18 @@ impl CompiledExpression {
                         if self.need_deref {
                             deref!();
                         }
+			print!("\n\n\nFinishing: translations {:?}?\n Buf: {:?}\n\n", old_to_new, code_buf);
+			for (from, to) in arcs {
+			    // relocate jump targets
+			    let new_from = old_to_new[from];
+			    let new_to = old_to_new[to];
+			    let new_diff = new_to as i32 - new_from as i32;
+                            code_buf[new_from - 2] = (new_diff & 0xFF) as u8;
```
Comment (ggreif) [inline crates/debug/src/transform/expression.rs:390]: This is hard-coded to be little-endian for now.

File: crates/debug/src/transform/expression.rs [hunk 3]
```diff
@@ -465,23 +576,31 @@ where
...
+	    [CompiledExpressionPart::LandingPad { original_pos }, CompiledExpressionPart::Code(code_chunk)]
+		if jump_arcs.values().any(|t| (original_pos + 1..original_pos + code_chunk.len()).contains(t))
+		=> Some(p[0].clone()),
+	    [CompiledExpressionPart::LandingPad { original_pos }, _]
+		if jump_arcs.iter().all(|(from, to)| from != original_pos && to != original_pos)
+		=> None,
+	    _ => Some(p[0].clone())
+	})
+	.collect();
+    Ok(Some((CompiledExpression { parts, need_deref }, jump_arcs)))
```
Comment (yurydelendik) [inline crates/debug/src/transform/expression.rs:603]: Let's make `jump_arcs` as field of `CompiledExpression` (`from_label` will have empty map)

File: crates/debug/src/transform/expression.rs [hunk 4]
```diff
@@ -376,9 +421,29 @@ pub fn compile_expression<R>(
...
     let buf = expr.0.to_slice()?;
     let mut parts = Vec::new();
+    macro_rules! push {
+        ($unread:expr, $part:expr) => {{
+            let part = $part;
+            if let (CompiledExpressionPart::Code(cc2), Some(CompiledExpressionPart::Code(cc1))) =
+                (&part, parts.last())
+            {
+                let mut combined = cc1.clone();
+                parts.pop();
```
Comment (yurydelendik) [inline crates/debug/src/transform/expression.rs:436]: is `combined` == to result of `parts.pop()`? so you don't have to clone

Comment (ggreif) [reply crates/debug/src/transform/expression.rs:436]: What I'd really like to do is:
``` rust
            if let (CompiledExpressionPart::Code(cc2), Some(CompiledExpressionPart::Code(cc1))) =
                (&part, parts.last())
            {
		cc1.extend_from_slice(cc2);
            } else {
                parts.push(CompiledExpressionPart::LandingPad {
                    original_pos: (expr.0.len().into_u64() - $unread) as usize,
                });
                parts.push(part)
            }
```
But for that `Code` needs to carry a mutable vector.
I tried various other ways of getting rid of the `.clone()` but I guess my Rust-fu is not strong enough for that. I'd love to see your idea written out though. Indeed the `.pop()` removes the `Code` which has `cc1` in it, and then it gets appended by `cc2` and. pushed back as a fresh `Code`.

**Comment by ggreif** [comment]
@yurydelendik Thanks for the review. I can't see where I have introduced O(n^2). The insertion of `LandingPad`s is linear in the number of parts, just like the removal of uninteresting ones. Post-translation, building of the `old_to_new` should be linear (modulo HashMap internals) and the  backpatching too.

I have contemplated about remembering the offset of every part that is pushed, but the parts are not 1:1 preserved in the transformation. With floating `LandingPad`s this is not a problem. Attaching an original offset to every part is also tricky, because the translation process would have trouble doing the bookkeeping of offsets.

Regarding `frame_base`, yes, this is a hole still, but I need a better understanding what `frame_base` is for. Is it for the location expressions that guide into structure members?

**Comment by yurydelendik** [comment]
> I can't see where I have introduced O(n^2).

It is near `filter_map(.. any()/all() ..).collect()` and some `old_to_new` operations.

>  I need a better understanding what `frame_base` is for

This expression is coming from `DW_AT_frame_base`. (Discard my idea about offsetting, I was incorrect there) It shall be inlined during each `DW_OP_fbreg` operation.

File: crates/debug/src/transform/expression.rs (hunk 4 above)
Comment (yurydelendik) [reply crates/debug/src/transform/expression.rs:436]: You are not wrong above, just use `parts.last_mut()` instead of `.last()`.

**Comment by yurydelendik** [comment]
> Attaching an original offset to every part is also tricky, because the translation process would have trouble doing the bookkeeping of offsets.

FWIW I am also thinking about having `parts: Vec<(CompiledExpressionPart, usize)>,` -- this will take care of bookkeeping. Though you will need to track parts last offset instead of first -- it is the same thing for this logic.

**Comment by ggreif** [comment]
Well, I get quadratic behaviour only when assuming # jumps are proportional to the # of `DW_OP`s. Similarly, `old_to_new` is bounded by the # of jumps times two. But in principle you are right.

Is it realistic that `DW_AT_frame_base` has jumps in it? Should be just bail if so?

**Comment by yurydelendik** [comment]
> Is it realistic that DW_AT_frame_base has jumps in it? Should we just bail if so?

Yes, let's assume DW_AT_frame_base has no jumps, otherwise let's bail.

**Comment by ggreif** [comment]
> FWIW I am also thinking about having `parts: Vec<(CompiledExpressionPart, usize)>,` -- this will take care of bookkeeping. Though you will need to track parts last offset instead of first -- it is the same thing for this logic.

I tried to go down that route, but got quickly lost in the `build_with_locals` bowels, string at `ranges_builder` and co.

File: crates/debug/src/transform/expression.rs (hunk 4 above)
Comment (ggreif) [reply crates/debug/src/transform/expression.rs:436]: Success!

File: crates/debug/src/transform/expression.rs [hunk 5]
```diff
@@ -727,7 +877,121 @@ mod tests {
...
+            CompiledExpression {
+                parts: vec![
+                    CompiledExpressionPart::Code(vec![49, 18]),
+                    CompiledExpressionPart::Jump {
+                        conditionally: true
+                    },
+                    CompiledExpressionPart::LandingPad { original_pos: 5 }, // capture from
+                    CompiledExpressionPart::Deref,
+                    CompiledExpressionPart::LandingPad { original_pos: 5 }, // capture to
+                    CompiledExpressionPart::Code(vec![6, 48, 159])
```
Comment (yurydelendik) [inline crates/debug/src/transform/expression.rs:968]: I think this and above test cases are incorrect. Can you verify that?

**Comment by yurydelendik** [comment]
I feel a little bit intimidated by the `old_to_new` logic here. I hope there is a chance you can adapt my proposal at https://github.com/yurydelendik/wasmtime/tree/ggreif/main with this PR. (If not I will still try my best but it is too hard)

File: crates/debug/src/transform/expression.rs (hunk 5 above)
Comment (ggreif) [reply crates/debug/src/transform/expression.rs:968]: The parsing tests pass
```
test transform::expression::tests::test_debug_parse_expressions ... ok
```
The tests themselves are not intended to make sense, if that is what you mean. The example in the PR description is an actual location expression with its translation.

**Comment by ggreif** [comment]
I am happy to look into your proposal (tomorrow, it is really late here)! Meanwhile the keys of `old_to_new` are the byte offsets for the `DW_OP`s before translation and the values in the map are the corresponding jump sources and targets of the in terms of the expanded location expression offsets. For every `Code` block preceded by a `LandingPad` I also add a few key->value mappings corresponding to the potential internal operation offsets. After the key->value map is fully built, I visit each `jump_arc` and look it up to obtain the translated jump arc. Since the source of the translated jump arc is directly behind the 3 byte `DW_OP_bra/skip`, we know how to backpatch the jump displacement.

File: crates/debug/src/transform/expression.rs (hunk 5 above)
Comment (ggreif) [reply crates/debug/src/transform/expression.rs:968]: The `original_pos: 5` appearing twice is probably which is confusing. It confused me too. But actually it makes sense. The jump goes from 5 to 7. The `LandingPad` before `Code` just orients it and tells that there are potential jump targets in `Code` (pre-translation) pos 6: `DW_OP_lit0`, pos 7: `DW_OP_stack_value`.

**Comment by ggreif** [comment]
I looked into your branch. You draw global ids from a static supply, which eliminates a bunch of offset complexity, and adds machinery in different other places. I still don't see where the pre-translation ids meet the post-translation ones. I'll look harder.

File: crates/debug/src/transform/expression.rs [hunk 6]
```diff
@@ -388,17 +459,40 @@ where
...
     }
-    let base_len = parts.len();
     let mut code_chunk = Vec::new();
     macro_rules! flush_code_chunk {
         () => {
             if !code_chunk.is_empty() {
-                parts.push(CompiledExpressionPart::Code(code_chunk));
-                code_chunk = Vec::new();
+                push!(CompiledExpressionPart::Code(code_chunk.clone()));
+                code_chunk.clear()
```
Comment (yurydelendik) [inline crates/debug/src/transform/expression.rs:467]: can we revert this change?

File: crates/debug/src/transform/expression.rs [hunk 7]
```diff
@@ -388,17 +459,40 @@ where
...
             }
         };
     };
+    // Find all landing pads by scanning bytes, do not care about
+    // false location at this moment.
+    // Looks hacky but it is fast and we may not need to be really exact.
```
Comment (yurydelendik) [inline crates/debug/src/transform/expression.rs:473]:
```suggestion
    // Looks hacky but it is fast; does not need to be really exact.
```

File: crates/debug/src/transform/expression.rs [hunk 8]
```diff
@@ -602,10 +744,49 @@ impl<'a, 'b> ValueLabelRangesBuilder<'a, 'b> {
...
 }
 
+/// Marker for tracking incoming jumps.
+/// Different when created new, and the same when cloned.
+#[derive(Clone, Eq)]
+struct JumpTargetMarker(Rc<u32>);
+
+impl JumpTargetMarker {
+    fn new() -> JumpTargetMarker {
+        // We need somewhat unique hash data -- using part of
```
Comment (yurydelendik) [inline crates/debug/src/transform/expression.rs:754]:
```suggestion
        // Create somewhat unique hash data -- using part of
```

File: crates/debug/src/transform/expression.rs [hunk 9]
```diff
@@ -630,15 +814,38 @@ mod tests {
...
     };
 
+    #[test]
+    fn test_debug_expression_jump_target() {
+        let m1 = JumpTargetMarker::new();
+        let m2 = JumpTargetMarker::new();
+        assert!(m1 != m2);
+        assert!(m1 == m1.clone());
+
+        // internal hash_data test
```
Comment (yurydelendik) [inline crates/debug/src/transform/expression.rs:843]:
```suggestion
        // Internal hash_data test (theoretically can fail intermittently).
```

File: crates/debug/src/transform/expression.rs (hunk 6 above)
Comment (ggreif) [reply crates/debug/src/transform/expression.rs:467]:
```suggestion
                parts.push(CompiledExpressionPart::Code(code_chunk));
                code_chunk = Vec::new();
```
like this?

Comment (ggreif) [inline crates/debug/src/transform/expression.rs:467]:
```suggestion
                parts.push(CompiledExpressionPart::Code(code_chunk));
                code_chunk = Vec::new();
```

Comment (yurydelendik) [reply crates/debug/src/transform/expression.rs:467]: Can `push!(` stay? I was thinking about `code_chunk.clone()` and `code_chunk = Vec::new()` stuff.

Comment (ggreif) [reply crates/debug/src/transform/expression.rs:467]: I hope to keep `push!(` (see above), but no idea how to fix the warning without artificially reading the `code_chunk` after the last macro invocation.

Comment (yurydelendik) [reply crates/debug/src/transform/expression.rs:467]: I see. Yeah, it just needs `if !code_chunk.is_empty() {` without `code_chunk = Vec::new()`

Comment (ggreif) [reply crates/debug/src/transform/expression.rs]: I did something that works. But I feel dense and am not sure if this is what you meant. Please feel free to fix.

File: crates/debug/src/transform/expression.rs [hunk 10]
```diff
@@ -388,17 +459,41 @@ where
...
+    // Find all landing pads by scanning bytes, do not care about
+    // false location at this moment.
+    // Looks hacky but it is fast; does not need to be really exact.
+    for i in 0..buf.len() - 2 {
+        let op = buf[i];
+        if op == gimli::constants::DW_OP_bra.0 || op == gimli::constants::DW_OP_skip.0 {
+            // TODO fix for big-endian
+            let offset = i16::from_le_bytes([buf[i + 1], buf[i + 2]]);
+            let origin = i + 3;
+            // Discarding out-of-bounds jumps (also some of falsly detected ops)
```
Comment (ggreif) [inline crates/debug/src/transform/expression.rs:481]:
```suggestion
            // Discarding out-of-bounds jumps (also some of falsely detected ops)
```

File: crates/debug/src/transform/expression.rs [hunk 11]
```diff
@@ -465,7 +465,7 @@ where
             if !code_chunk.is_empty() {
                 push!(CompiledExpressionPart::Code(code_chunk));
                 code_chunk = Vec::new();
-                if code_chunk.is_empty() {} // suppresses warning
+                let _ = code_chunk; // suppresses warning for final flush
```
Comment (ggreif) [inline crates/debug/src/transform/expression.rs:468]: I see, thanks!

**Comment by yurydelendik** [comment]
Thank you for the patch
